internal/kv/etcd/etcd_kv.go

```diff
@@ -32,12 +36,116 @@ type EtcdKV struct {
 	rootPath string
 }
 
+const (
+	kCfgKeyTlsBase   = "etcd.tls"
+	kCfgKeyTlsCert   = kCfgKeyTlsBase + ".cert"
+	kCfgKeyTlsKey    = kCfgKeyTlsBase + ".key"
+	kCfgKeyTlsCACert = kCfgKeyTlsBase + ".caCert"
+)
+
+func loadParam(table paramTableGetter, key string) (string, error) {
+	val, err := table.Load(key)
+	if err != nil {
+		return "", errors.Wrapf(err, "get parameter from global paramtable, key=%s", key)
+	}
+	return val, nil
+}
+
+func loadParamAndEnsureNotEmpty(table paramTableGetter, key string) (string, error) {
+	v, err := loadParam(table, key)
+	if err != nil {
+		return "", err
+	}
+	if len(v) == 0 {
+		return "", errors.Errorf("field %s must set in the global paramtable", key)
+	}
+	return v, nil
+}
+
+func newEtcdClient(endpoints []string, table paramTableGetter) (*clientv3.Client, error) {
+	certFile, err := loadParam(table, kCfgKeyTlsCert)
+	if err != nil {
+		return nil, err
+	}
+	var cfg clientv3.Config
+	cfg.Endpoints = endpoints
+	cfg.DialTimeout = 5 * time.Second
+
+	// if certFile is set, lookup other fields
+	if len(certFile) != 0 {
```

Comment:
I don't think so. 

The parameter table should be just a container for configuration fields and should not be coupled with the factory methods of instance.  Otherwise, the parameter table will be changed frequently.

I think it would be better to use an `IoC` container and make the configuration as a global k-v (or even just add *viper.Viper into this container.) Each component just parses configuration all by itself.

Like

```go
func RegisterSelfIntoIoCContainer(container* ioc.Container) {
  container.RegisterSingleton(func (viper *viper.Viper) (EtcdKV, error) {
    var config struct {
        Endpoint []string
        TLS struct {
            ...
        }
    }
    etcdConfig = viper.Get("etcd")
    err := mapstructure.Decode(etcdConfig, &config)
    ...
  })
}
```

In this design, all global variables, `ParamTables`, `NewEtcdKV` calls, will be extinguished and the code will be MUCH MUCH cleaner since each component does its own business well.


---

避免英语有歧义，我再用中文转述一遍。

`ParamTable`是一个很不好的设计。因为，如果ParamTable要耦合所有的模块的各种配置类型，那相当于只要有新的模块或者新的配置类型，`ParamTable`都要修改，也相当于不同的进程都要手写一遍`ParamTable`的全局变量。

所以道理上，一个模块如何用配置文件，应该是这个模块自己的事情，而不是一个全局类型。

最简单的其实如果能用IoC容器的话，将所有配置文件都解析到一个 `*viper.Viper` 里。然后，EtcdKV等，都是从容器里拿出 *viper.Viper 然后生成 etcd kv就好了。同时，全局时钟(TSO) 也可以从容器内拿出来etcdkv和*viper.Viper，自己初始化。

本质上，一个模块自己的事情应该自己干。如何初始化一个模块，应该是这个模块自己的事儿。实现这个事儿的手段就是IoC容器。

比较简单的golang ioc容器，可以参考 https://github.com/golobby/container

Reply:
Agree current paramtable is a bad design. 

As you said, I want to refactor param table so viper load same configuration files for each module ,like a global k-v.
All the component see the same configurations and use only what they care about.

On the other hand,  if we import golang ioc and each component share the same singleton that would be great. 

The only concern about this patch is that there are a lot a config parse logic in etcd kv, which I believe should be a part for paramUtil or some helper function but not in etcdkv it self.

We can start to refactor some of the part you mentioned on later issues